cloudwatch_logs_common/src/log_publisher.cpp

```diff
@@ -108,102 +107,153 @@ Aws::CloudWatchLogs::ROSCloudWatchLogsErrors LogPublisher::StopPublisherThread()
+ */
+void LogPublisher::SendLogs(Aws::String & next_token)
+{
+  Aws::CloudWatchLogs::ROSCloudWatchLogsErrors publisher_status = CW_LOGS_SUCCEEDED;
+  auto shared_logs_obj = this->shared_logs_.load(std::memory_order_acquire);
+  if (nullptr != shared_logs_obj) {
+    std::list<Aws::CloudWatchLogs::Model::InputLogEvent> * logs;
+    publisher_status = shared_logs_obj->getDataAndLock(logs);
+    if (CW_LOGS_SUCCEEDED == publisher_status) {
+      if (!logs->empty()) {
+        Aws::CloudWatchLogs::ROSCloudWatchLogsErrors send_logs_status = CW_LOGS_FAILED;
+        int tries = kMaxRetries;
+
+        while (CW_LOGS_SUCCEEDED != send_logs_status && tries > 0) {
+          send_logs_status = this->cloudwatch_facade_->SendLogsToCloudWatch(
+            next_token, this->log_group_, this->log_stream_, logs);
           if (CW_LOGS_SUCCEEDED != send_logs_status) {
-            AWS_LOG_WARN(
-              __func__,
-              "Unable to send logs to CloudWatch and retried, dropping this batch of logs ...");
+            AWS_LOG_WARN(__func__, "Unable to send logs to CloudWatch, retrying ...");
+            Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
+              this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
+                                                          next_token);
+            if (CW_LOGS_SUCCEEDED != get_token_status) {
+              AWS_LOG_WARN(__func__,
+                           "Unable to obtain the sequence token to use, quit attempting to send "
+                           "logs to CloudWatch");
+              break;
+            }
           }
+          tries--;
+        }
+        if (CW_LOGS_SUCCEEDED != send_logs_status) {
+          AWS_LOG_WARN(
+            __func__,
+            "Unable to send logs to CloudWatch and retried, dropping this batch of logs ...");
         }
       }
-      // TODO: For now we're just going to discard logs that fail to get sent. Later we may add some
-      // retry strategy
-
-      /* Null out the class reference before unlocking the object. The external thread will be
-         looking to key off if the shared object is locked or not to know it can start again, but
-         the PublishLogs function checks if the the pointer is null or not. */
-      this->shared_logs_.store(nullptr, std::memory_order_release);
-      shared_logs_obj->freeDataAndUnlock();
     }
-    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    // TODO: For now we're just going to discard logs that fail to get sent. Later we may add some
+    // retry strategy
+
+    /* Null out the class reference before unlocking the object. The external thread will be
+       looking to key off if the shared object is locked or not to know it can start again, but
+       the PublishLogs function checks if the the pointer is null or not. */
+    this->shared_logs_.store(nullptr, std::memory_order_release);
+    shared_logs_obj->freeDataAndUnlock();
```

Comment:
This is unrelated but this `getDataAndLock` / `freeDataAndUnlock` is a bit hard to understand.
Here is a proposal for a ThreadSafeStack. This class could use a stack of size 1. Design can always be simplified if necessary to hold a single element.

```c++
#include <array>
#include <exception>
#include <memory>
#include <mutex>

class StackEmpty : public std::exception {
  const char * what() const;
};

class StackFull : public std::exception {
  const char * what() const;
};

// TODO: this class should be annotated with TSAs.
template <typename T, int N>
class ThreadSafeStack final {
public:
  constexpr ThreadSafeStack() = default;  // can be constexpr because std::array, unique_ptr, int and std::mutex all have constexpr constructors.
  ~ThreadSafeStack() noexcept = default;

  ThreadSafeArray(const ThreadSafeArray &) = delete;
  ThreadSafeArray operator=(const ThreadSafeArray &) = delete;

  void push(std::unique_ptr<T> &&new_element) {
    bool stack_is_full = false;
    {
      std::scoped_lock<std::mutex> lock(mutex_);
      if (size_ == N) {
        stack_is_full = true;
      } else {
        array_[size++] = std::move(new_element);
      }
    }
    if (stack_is_full) {
      throw StackFull();
  }

  std::unique_ptr<T> pop() {
    bool stack_is_empty = false;
    std::unique_ptr popped_element;
    {
      std::scoped_lock<std::mutex> lock(mutex_);
      if (!!size_) {
        stack_is_empty = true;
      } else {
        popped_element = std::move(array_[size_--]);
      }
    }
    if (stack_is_empty) {
      throw StackEmpty();
    }
    return popped_element;
  }

  bool empty() const {
    std::scoped_lock<std::mutex> lock(mutex_);
    return !!size_;
  }

  int size() const {
    std::scoped_lock<std::mutex> lock(mutex_);
    return size_;
  }

  constexpr int max_size() const {
    return N;
   }

private:
  std::array<std::unique_ptr<T>, N> array_{};
  int size_{0};
  mutable std::mutex mutex_{};
};
```

Note: this has been inspired by the Chapter 6 of [C++ Concurrency in Action](https://www.amazon.com/C-Concurrency-Action-Practical-Multithreading/dp/1933988770).

A few things to notice:

* Restricting this class to using `unique_ptr`  is great because it makes the memory ownership unambiguous. `push()` caller owns the element, then the ownership is transferred to the container, then transferred to the `pop()` caller. This avoids tricky cases where the initial thread keeps a reference on an element, it gets popped by another thread and we start to need caring about `std::shared_ptr` [thread-safety properties](https://stackoverflow.com/questions/14482830/stdshared-ptr-thread-safety). Guaranteeing the change of ownership is possible by passing the element to `push()` as an RR-value.
* [No lock :tada: ](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-lockfree)
* No memory allocation
* STL-like API

This is already a good step in the right direction, but the ideal data structure we want is really a FIFO container. C++ concurrency in action, Chapter 6, Listing 6.9 is very close to what we want (a circular buffer would be better to avoid dynamic memory allocation). Compare to this design, it properly uses condition variable to implement `wait_and_pop` which removes the need for sleeping in the publisher thread.

Reply:
`shared_logs_` is a reference 

> This code is hard to understand. Why is `this->shared_logs_` and `shared_logs_obj` both used?

The `this->shared_logs_` is an atomic holding a reference to a shared buffer and `shared_logs_obj` is the copy of that reference. The atomic is used to pass a reference to a shared buffer between the two threads safely. We clear the atomic because we don't need the reference anymore and we free the shared object to notify the producer thread that it can start using it again and hand us the next buffer. 

Like I mentioned elsewhere, it was a premature optimization and can probably be simplified during the next iteration of development when we have better requirements.